lib/Component.spec/getDerivedStateFromError.spec.lua

```diff
@@ -0,0 +1,190 @@
+return function()
+    local createElement = require(script.Parent.Parent.createElement)
+    local createReconciler = require(script.Parent.Parent.createReconciler)
+    local NoopRenderer = require(script.Parent.Parent.NoopRenderer)
+
+    local Component = require(script.Parent.Parent.Component)
+
+    local noopReconciler = createReconciler(NoopRenderer)
+
+    it("should not be called if the component doesn't throw", function()
+        local callCount = 0
+
+        local Boundary = Component:extend("Boundary")
+
+        function Boundary.getDerivedStateFromError(message)
+            callCount = callCount + 1
+        end
+
+        function Boundary:render()
+            return nil
+        end
+
+        local element = createElement(Boundary)
+        local hostParent = nil
+        local key = "Test"
+
+        noopReconciler.mountVirtualNode(element, hostParent, key)
+        expect(callCount).to.equal(0)
+    end)
+
+    it("should be called with the error message", function()
+        local callCount = 0
+
+        local function Bug()
+            error("test error")
+        end
+
+        local Boundary = Component:extend("Boundary")
+
+        function Boundary.getDerivedStateFromError(message)
+            callCount = callCount + 1
+            -- the error message will not be the same as what's thrown because a
+            -- line/source object as well as stack trace will be attached to it
+            expect(message).to.be.ok()
+
+            return {}
+        end
+
+        function Boundary:render()
+            return createElement(Bug)
+        end
+
+        local element = createElement(Boundary)
+        local hostParent = nil
+        local key = "Test"
+
+        -- This will throw, because we don't stop rendering the throwing
+        -- component in response to the error. We test this more
+        -- rigorously elsewhere; this is just to keep the test from failing.
+        expect(function()
+            noopReconciler.mountVirtualNode(element, hostParent, key)
+        end).to.throw()
+        expect(callCount).to.equal(1)
+    end)
```

Comment:
You should use `createSpy` for these tests! I think it might reduce a lot of your boilerplate, especially if you use `spy.assertCalledWith`

Reply:
Yeah! `assertCalledWith` isn't super useful here but I can use it in #174.